lib/tapioca/compilers/dynamic_mixin_compiler.rb

```diff
@@ -0,0 +1,199 @@
+
+  sig { returns(T::Array[Symbol]) }
+  attr_reader :instance_attribute_readers, :instance_attribute_writers, :instance_attribute_predicates
+
+  sig { params(constant: Module).void }
+  def initialize(constant)
+    @constant = constant
+    mixins_from_modules = {}.compare_by_identity
+    class_attribute_readers = T.let([], T::Array[Symbol])
+    class_attribute_writers = T.let([], T::Array[Symbol])
+    class_attribute_predicates = T.let([], T::Array[Symbol])
+
+    instance_attribute_readers = T.let([], T::Array[Symbol])
+    instance_attribute_writers = T.let([], T::Array[Symbol])
+    instance_attribute_predicates = T.let([], T::Array[Symbol])
+
+    Class.new do
+      # Override the `self.include` method
+      define_singleton_method(:include) do |mod|
+        # Take a snapshot of the list of singleton class ancestors
+        # before the actual include
+        before = singleton_class.ancestors
+        # Call the actual `include` method with the supplied module
+        super(mod).tap do
+          # Take a snapshot of the list of singleton class ancestors
+          # after the actual include
+          after = singleton_class.ancestors
+          # The difference is the modules that are added to the list
+          # of ancestors of the singleton class. Those are all the
+          # modules that were `extend`ed due to the `include` call.
+          #
+          # We record those modules on our lookup table keyed by
+          # the included module with the values being all the modules
+          # that that module pulls into the singleton class.
+          #
+          # We need to reverse the order, since the extend order should
+          # be the inverse of the ancestor order. That is, earlier
+          # extended modules would be later in the ancestor chain.
+          mixins_from_modules[mod] = (after - before).reverse!
+        end
+      rescue Exception # rubocop:disable Lint/RescueException
+        # this is a best effort, bail if we can't perform this
+      end
+
+      define_singleton_method(:class_attribute) do |*attrs, **kwargs|
+        class_attribute_readers.concat(attrs)
+        class_attribute_writers.concat(attrs)
+
+        instance_predicate = kwargs.fetch(:instance_predicate, true)
+        instance_accessor = kwargs.fetch(:instance_accessor, true)
+        instance_reader = kwargs.fetch(:instance_reader, instance_accessor)
+        instance_writer = kwargs.fetch(:instance_writer, instance_accessor)
+
+        if instance_reader
+          instance_attribute_readers.concat(attrs)
+        end
+
+        if instance_writer
+          instance_attribute_writers.concat(attrs)
+        end
+
+        if instance_predicate
+          class_attribute_predicates.concat(attrs)
+
+          if instance_reader
+            instance_attribute_predicates.concat(attrs)
+          end
+        end
+
+        super(*attrs, **kwargs) if defined?(super)
```

Comment:
Can't we just do:
```suggestion
        super if defined?(super)
```
to pass all the arguments implicitly?

Reply:
Hmmm, yeah, it could work differently indeed.